deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/resource/EnvDeploys.java

```diff
@@ -227,13 +227,24 @@ public Response create(
                     buildBean.getArtifact_url()));
         }
         // if the stage is not allowed (allow_private_build)
-        if(! envBean.getAllow_private_build()) { 
+        if(! envBean.getAllow_private_build()) {
             // only allow deploy if it is not private build
             if (buildBean.getScm_branch().equals("private")) {
                 throw new TeletaanInternalException(Response.Status.BAD_REQUEST, 
                     "This stage does not allow deploying a private build. Please Contact #teletraan to allow your stage for deploying private build");
             }
         }
+        // disallow sox deploy if the build artifact is private
+        if(envBean.getIs_sox() && buildBean.getScm_branch().equals("private")) {
+            throw new TeletaanInternalException(Response.Status.BAD_REQUEST,
+                "This stage requires SOX builds. A private build cannot be used in a sox-compliant stage.");
+        }
+        // disallow sox deploy if the build artifact is not from a sox source url
+        if(envBean.getIs_sox() && !buildAllowlist.sox_compliant(buildBean.getArtifact_url())) {
+            throw new TeletaanInternalException(Response.Status.BAD_REQUEST,
+                String.format("This stage requires SOX builds. The build must must be from a sox-compliant location (%s). Please Contact #teletraan to ensure the build artifact is published to a sox-compliant url",
```

Comment:
remove a part of the message referencing to channel. We can direct users on next action instead. ex: print list of allowed sox urls